src/charm.py

```diff
@@ -0,0 +1,160 @@
+from base64 import b64encode
+
+import yaml
+from charmed_kubeflow_chisme.exceptions import ErrorWithStatus
+from charmed_kubeflow_chisme.kubernetes import KubernetesResourceHandler
+from lightkube import ApiError
+from lightkube.generic_resource import load_in_cluster_generic_resources
+from ops.charm import CharmBase
+from ops.framework import StoredState
+from ops.main import main
+from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus, WaitingStatus
+
+from certs import gen_certs
+
+K8S_RESOURCE_FILES = ["src/templates/webhook_resources.yaml"]
+TAGGED_IMAGE = "charmedkubeflow/namespace-node-affinity:90dde45ab265af91369d09a377a26034bc453a5d"
+
+
+class NamespaceNodeAffinityOperator(CharmBase):
+    """A Juju Charm for Namespace Node Affinity."""
+
+    _stored = StoredState()
+
+    def __init__(self, *args):
+        """Initialize charm."""
+        super().__init__(*args)
+
+        # convenience variables and base settings
+        self.logger = logging.getLogger(__name__)
+        self._namespace = self.model.name
+        self._lightkube_field_manager = "lightkube"
+        self._name = self.model.app.name
+
+        self._gen_certs_if_missing()
+
+        self._k8s_resource_handler = None
+
+        # setup events
+        self.framework.observe(self.on.config_changed, self.main)
+        self.framework.observe(self.on.install, self.main)
+        self.framework.observe(self.on.leader_elected, self.main)
+        self.framework.observe(self.on.remove, self.main)
+        self.framework.observe(self.on.upgrade_charm, self.main)
+
+    def main(self, _):
+        """Entrypoint for most charm events."""
+        self.logger.info("Starting main")
+        try:
+            self._check_leader()
+            self._deploy_k8s_resources()
+        except ErrorWithStatus as error:
+            self.model.unit.status = error.status
+            return
+
+        self.model.unit.status = ActiveStatus()
+
+    def _check_leader(self):
+        """Check if this unit is a leader."""
+        self.logger.info("_check_leader")
+        if not self.unit.is_leader():
+            self.logger.info("Not a leader, skipping setup")
+            raise ErrorWithStatus("Waiting for leadership", WaitingStatus)
+
+    def _deploy_k8s_resources(self) -> None:
+        """Deploy K8S resources."""
+        self.logger.info("_deploy_k8s_resources")
+        try:
+            self.unit.status = MaintenanceStatus("Creating K8S resources")
+            self.k8s_resource_handler.apply()
+        except ApiError:
+            self.logger.error("K8S resource creation failed with ApiError:")
+            self.logger.error(str(ApiError))
+            self.logger.error(ApiError.status)
+            raise ErrorWithStatus("K8S resources creation failed", BlockedStatus)
+
+        self.model.unit.status = MaintenanceStatus("K8S resources created")
+
+    def _gen_certs_if_missing(self):
+        """Generate certificates if they don't already exist in _stored."""
```

Comment:
Just a comment. When this is in place, certificates will be only generated on initial install. What about upgrade? Do we want to re-generate certificates in such case?

Reply:
Good question.  The intent was to avoid cert-churn (replacing certs every time the charm fires, even when there already are certs).  Someone brought that up as a problem a while ago for some other charm (katib maybe?).  But maybe I've hit the wrong balance

Do you have thoughts on when we'd want to recreate certs?  Initially I thought maybe we just want to recreate certs whenever they're stale/dont exist, and continue to use existing ones otherwise.  Although now that I think about it, if this charm runs for a year (the lifetime of those certs) it has no mechanism for refreshing them does it?

Reply:
Yes, that was my concern: there is no way to re-fresh certificates.

Reply:
But on other hand should upgrade re-generate certificates? May be we should generate when they don't exist (as it is right now), but also re-generate in upgrade?
We can solve re-generation of certificates by restarting pods, right?